presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,550 @@
+        boolean includeNulls = nullsAllowed && presentStream != null;
+        if (positionCount != outputPositionCount) {
+            compactValues(positions, positionCount, includeNulls, nulls);
+        }
+        return newLease(new VariableWidthBlock(positionCount, dataAsSlice, offsets, Optional.ofNullable(includeNulls ? nulls : null)));
+    }
+
+    private BlockLease newLease(Block block)
+    {
+        valuesInUse = true;
+        return ClosingBlockLease.newLease(block, () -> valuesInUse = false);
+    }
+
+    @Override
+    public void throwAnyError(int[] positions, int positionCount)
+    {
+    }
+
+    @Override
+    public void close()
+    {
+        systemMemoryContext.close();
+    }
+
+    private void openRowGroup()
+            throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+        lengthStream = lengthStreamSource.openStream();
+        dataStream = dataStreamSource.openStream();
+
+        rowGroupOpen = true;
+    }
+
+    @Override
+    public void startStripe(InputStreamSources dictionaryStreamSources, List<ColumnEncoding> encoding)
+    {
+        presentStreamSource = missingStreamSource(BooleanInputStream.class);
+        lengthStreamSource = missingStreamSource(LongInputStream.class);
+        dataStreamSource = missingStreamSource(ByteArrayInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        lengthStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public void startRowGroup(InputStreamSources dataStreamSources)
+    {
+        presentStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, PRESENT, BooleanInputStream.class);
+        lengthStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, LENGTH, LongInputStream.class);
+        dataStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, DATA, ByteArrayInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        lengthStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + sizeOf(offsets) + sizeOf(outputPositions) + sizeOf(data) + sizeOf(nulls) + sizeOf(lengths) + sizeOf(isNullVector);
+    }
+
+    private void prepareForNextRead(int positionCount, int[] positions)
+            throws IOException
+    {
+        lengthIndex = 0;
+        outputPositionCount = 0;
+
+        int totalLength = 0;
+        int maxLength = 0;
+
+        int totalPositions = positions[positionCount - 1] + 1;
+        int nullCount = 0;
+        if (presentStream != null) {
+            isNullVector = ensureCapacity(isNullVector, totalPositions);
+            nullCount = presentStream.getUnsetBits(totalPositions, isNullVector);
+        }
+
+        if (lengthStream != null) {
+            lengths = ensureCapacity(lengths, totalPositions - nullCount);
```

Comment:
nit: consider extracting totalPositions - nullCount into a local variable and reusing in 3 places: `int nonNullCount = totalPositions - nullCount;`